packages/calendar-input/src/date-input.tsx

```diff
@@ -0,0 +1,274 @@
+import * as React from 'react'
+import * as S from './styles'
+import {
+  CalendarDate as CalendarDateHelper,
+  CalendarDateTime,
+  DateTimePartsType,
+  DateValue,
+  KEYS
+} from './utils'
+import {CalendarIcon, ChevronUpIcon} from './icons'
+import {DateInputProps} from './types'
+
+export const DateInput: React.FC<DateInputProps> = ({
+  onChange = () => {},
+  onClick,
+  value,
+  getRef,
+  openCalendar,
+  withTime = false,
+  minutesInterval,
+  timeValue,
+  isExpanded,
+  expandIcon = () => <CalendarIcon />,
+  collapseIcon = () => <ChevronUpIcon />,
+  ...props
+}) => {
+  const CalendarDate = withTime
+    ? new CalendarDateTime()
+    : new CalendarDateHelper()
+  const [inputValue, setInputValue] = React.useState<DateValue>(
+    CalendarDate.initValue
+  )
+  const inputNode = React.useRef<HTMLInputElement | null>(null)
+  const [cursorPosition, setCursorPosition] = React.useState(0)
+  const [activeDatePart, setActiveDatePart] = React.useState<Partial<
+    DateTimePartsType
+  > | null>(null)
+
+  React.useEffect(() => {
+    setSelectionRange(cursorPosition, cursorPosition + 1)
+  }, [cursorPosition])
+  React.useEffect(() => {
+    if (value) {
+      setInputValue(CalendarDate.getValue(value))
+    } else {
+      setInputValue({
+        ...inputValue,
+        hours: timeValue.hours || CalendarDate.initValue.hours,
+        minutes: timeValue.minutes || CalendarDate.initValue.minutes
+      })
+    }
+  }, [value, timeValue])
+
+  const changeInputValue = (newValue: DateValue) => {
+    const isValidDate = !Object.values(newValue).some(elem =>
+      isNaN(Number(elem))
+    )
+    if (isValidDate) {
+      const minutes = Number(newValue.minutes)
+      const increaseMinutes = minutes % minutesInterval > minutesInterval / 2
+      onChange(
+        CalendarDate.getDate({
+          ...newValue,
+          minutes:
+            minutes -
+            (minutes % minutesInterval) +
+            (increaseMinutes ? minutesInterval : 0)
+        })
+      )
+    } else {
+      onChange()
+      setInputValue(newValue)
+    }
+  }
+
+  const setSelectionRange = (start: number, end: number) => {
+    if (inputNode.current) {
+      inputNode.current.childNodes.forEach(node => {
+        if (node.nodeName === 'INPUT' && activeDatePart) {
+          ;(node as HTMLInputElement).focus()
+          ;(node as HTMLInputElement).setSelectionRange(start, end)
+        }
+      })
+    }
+  }
+
+  const focusNextPart = () => {
```

Comment:
Would it be possible to traverse input parts using tab?

Reply:
Yes, you can do it now

Reply:
![](https://i.imgur.com/IazJWlU.jpg)